assets/scripts/segments/drag_and_drop.js

```diff
@@ -667,118 +355,272 @@ function doDropHeuristics (type, variantString, width) {
+    const hoverIndex = props.dataNo
 
-  changeDraggingType(DRAGGING_TYPE_NONE)
+    const hoveredSegment = component.getDecoratedComponentInstance().streetSegment
+    const { left } = hoveredSegment.getBoundingClientRect()
+    const hoverMiddleX = Math.round(left + (props.width) / 2)
+    const { x } = monitor.getClientOffset()
 
-  if (segmentElControls) {
-    scheduleControlsFadeout(segmentElControls)
+    if (dragIndex === hoverIndex) {
+      updateIfDraggingStateChanged(undefined, dragIndex, monitor.getItem())
+    } else {
+      const { segments } = store.getState().street
+
+      const segmentBeforeEl = (x > hoverMiddleX && hoverIndex !== segments.length - 1) ? hoverIndex + 1
+        : (hoverIndex === segments.length - 1) ? undefined
+          : hoverIndex
+
+      const segmentAfterEl = (x > hoverMiddleX && hoverIndex !== 0) ? hoverIndex
+        : (hoverIndex === 0) ? undefined
+          : hoverIndex - 1
+
+      updateIfDraggingStateChanged(segmentBeforeEl, segmentAfterEl, monitor.getItem())
+    }
   }
+}
+
+function handleSegmentCanvasDrop (draggedItem) {
+  const { segmentBeforeEl, segmentAfterEl, draggedSegment } = store.getState().ui.draggingState
+
+  store.dispatch(clearDraggingState())
+  // If dropped in same position as dragged segment was before, return
+  if (segmentBeforeEl === undefined && segmentAfterEl === draggedItem.dataNo) {
+    store.dispatch(setActiveSegment(draggedItem.dataNo))
+    return
+  }
+
+  const newSegment = {
+    variantString: draggedItem.variantString,
+    width: draggedItem.width / TILE_SIZE,
+    type: draggedItem.type,
+    randSeed: draggedItem.randSeed
+  }
+
+  let newIndex = (segmentAfterEl !== undefined) ? (segmentAfterEl + 1) : segmentBeforeEl
 
-  if (failedDrop) {
-    infoBubble.show(true)
+  if (!draggedItem.forPalette) {
+    store.dispatch(removeSegment(draggedSegment))
+    newIndex = (newIndex < draggedSegment) ? newIndex : newIndex - 1
   }
+
+  store.dispatch(addSegment(newIndex, newSegment))
+  store.dispatch(setActiveSegment(newIndex))
 }
 
-export function onBodyMouseUp (event) {
-  switch (_draggingType) {
-    case DRAGGING_TYPE_NONE:
-      return
-    case DRAGGING_TYPE_CLICK_OR_MOVE:
-      changeDraggingType(DRAGGING_TYPE_NONE)
-      setIgnoreStreetChanges(false)
-      break
-    case DRAGGING_TYPE_MOVE:
-      handleSegmentMoveEnd(event)
-      break
-    case DRAGGING_TYPE_RESIZE:
-      handleSegmentResizeEnd(event)
-      break
+/**
+ * Determines if segment was dropped/hovered on left or right side of street
+ *
+ * @param {Node} segment - reference to StreetEditable
+ * @param {Number} droppedPosition - x position of dropped segment in reference
+ *    to StreetEditable
+ * @returns {string} - left, right, or null if dropped/hovered over a segment
+ */
+function isOverLeftOrRightCanvas (segment, droppedPosition) {
+  const { remainingWidth } = store.getState().street
+  const { left, right } = segment.getBoundingClientRect()
```

Comment:
I'm seeing that sometimes the `segment` is undefined during a drag action, which causes the dragged item to be frozen in place, then lost once I let go of the drag.

Reply:
![jul-09-2018 14-35-18](https://user-images.githubusercontent.com/2553268/42469323-5f9cd5ea-8385-11e8-8461-06fb4dc11034.gif)

Reply:
Never mind, I think this was my fault, I was trying something weird. IGNORE!